Assertion '_toolbar == null' failure in TextField double-tap

## Steps to Reproduce
We received this error stack trace from a crash report.

## Logs
```
Thread 0 main thread_AssertionError: 'package:flutter/src/widgets/text_selection.dart': Failed assertion: line 357 pos 12: '_toolbar == null': is not true.
at _AssertionError._doThrowNew	(errors_patch.dart:40 )
at _AssertionError._throwNew	(errors_patch.dart:36 )
at TextSelectionOverlay.showToolbar	(text_selection.dart:357 )
at EditableTextState.showToolbar	(editable_text.dart:1393 )
at _TextFieldState._handleDoubleTapDown	(text_field.dart:800 )
at _TextSelectionGestureDetectorState._handleTapDown	(text_selection.dart:821 )
at TapGestureRecognizer._checkDown.<anonymous closure>	(tap.dart:328 )
at GestureRecognizer.invokeCallback	(recognizer.dart:181 )
at TapGestureRecognizer._checkDown	(tap.dart:328 )
at TapGestureRecognizer.didExceedDeadlineWithEvent	(tap.dart:292 )
at PrimaryPointerGestureRecognizer.addAllowedPointer.<anonymous closure>	(recognizer.dart:429 )
at _rootRun	(zone.dart:1120 )
at _CustomZone.run	(zone.dart:1021 )
at _CustomZone.runGuarded	(zone.dart:923 )
at _CustomZone.bindCallbackGuarded.<anonymous closure>	(zone.dart:963 )
at _rootRun	(zone.dart:1124 )
at _CustomZone.run	(zone.dart:1021 )
at _CustomZone.bindCallback.<anonymous closure>	(zone.dart:947 )
at Timer._createTimer.<anonymous closure>	(timer_patch.dart:21 )
at _Timer._runTimers	(timer_impl.dart:382 )
at _Timer._handleMessage	(timer_impl.dart:416 )
at _RawReceivePortImpl._handleMessage	(isolate_patch.dart:171 )
```

--- a/packages/flutter/lib/src/material/text_field.dart
+++ b/packages/flutter/lib/src/material/text_field.dart
@@ -721,8 +721,9 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
         from: details.globalPosition,
         cause: SelectionChangedCause.forcePress,
       );
-      if (_shouldShowSelectionToolbar)
+      if (_shouldShowSelectionToolbar) {
         _editableTextKey.currentState.showToolbar();
+      }
     }
   }
 
@@ -796,8 +797,9 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
   void _handleDoubleTapDown(TapDownDetails details) {
     if (widget.selectionEnabled) {
       _renderEditable.selectWord(cause: SelectionChangedCause.doubleTap);
-      if (_shouldShowSelectionToolbar)
-        _editableTextKey.currentState.showToolbar();
+      if (_shouldShowSelectionToolbar) {
+        _editableText.toggleToolbar();
+      }
     }
   }
 

--- a/packages/flutter/lib/src/material/text_field.dart
+++ b/packages/flutter/lib/src/material/text_field.dart
@@ -722,7 +722,7 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
         cause: SelectionChangedCause.forcePress,
       );
       if (_shouldShowSelectionToolbar) {
-        _editableTextKey.currentState.showToolbar();
+        _editableTextKey.currentState.toggleToolbar();
       }
     }
   }
@@ -790,8 +790,10 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
   }
 
   void _handleSingleLongTapEnd(LongPressEndDetails details) {
-    if (_shouldShowSelectionToolbar)
-      _editableTextKey.currentState.showToolbar();
+    if (widget.selectionEnabled) {
+      if (_shouldShowSelectionToolbar)
+        _editableTextKey.currentState.toggleToolbar();
+    }
   }
 
   void _handleDoubleTapDown(TapDownDetails details) {

--- a/packages/flutter/test/material/text_field_test.dart
+++ b/packages/flutter/test/material/text_field_test.dart
@@ -5035,6 +5035,36 @@ void main() {
   );
 
   testWidgets(
+    'double long press toggles selection menu',
+    (WidgetTester tester) async {
+      final TextEditingController controller = TextEditingController(
+        text: '',
+      );
+      await tester.pumpWidget(
+        MaterialApp(
+          home: Material(
+            child: Center(
+              child: TextField(
+                controller: controller,
+              ),
+            ),
+          ),
+        ),
+      );
+
+      // Long press shows the selection menu.
+      await tester.longPressAt(textOffsetToPosition(tester, 0));
+      await tester.pump();
+      expect(find.text('PASTE'), findsOneWidget);
+
+      // Double tap again hides the selection menu.
+      await tester.longPressAt(textOffsetToPosition(tester, 0));
+      await tester.pump();
+      expect(find.text('PASTE'), findsNWidgets(0));
+    },
+  );
+
+  testWidgets(
     'double tap hold selects word (iOS)',
     (WidgetTester tester) async {
       final TextEditingController controller = TextEditingController(

--- a/packages/flutter/test/material/text_field_test.dart
+++ b/packages/flutter/test/material/text_field_test.dart
@@ -5001,6 +5001,40 @@ void main() {
   );
 
   testWidgets(
+    'double double tap toggles selection menu',
+    (WidgetTester tester) async {
+      final TextEditingController controller = TextEditingController(
+        text: '',
+      );
+      await tester.pumpWidget(
+        MaterialApp(
+          home: Material(
+            child: Center(
+              child: TextField(
+                controller: controller,
+              ),
+            ),
+          ),
+        ),
+      );
+
+      // Double tap on the same location shows the selection menu.
+      await tester.tapAt(textOffsetToPosition(tester, 0));
+      await tester.pump(const Duration(milliseconds: 50));
+      await tester.tapAt(textOffsetToPosition(tester, 0));
+      await tester.pump();
+      expect(find.text('PASTE'), findsOneWidget);
+
+      // Double tap again hides the selection menu.
+      await tester.tapAt(textOffsetToPosition(tester, 0));
+      await tester.pump(const Duration(milliseconds: 50));
+      await tester.tapAt(textOffsetToPosition(tester, 0));
+      await tester.pump();
+      expect(find.text('PASTE'), findsNWidgets(0));
+    },
+  );
+
+  testWidgets(
     'double tap hold selects word (iOS)',
     (WidgetTester tester) async {
       final TextEditingController controller = TextEditingController(

